In [60]:
import numpy as np
import pandas as pd

from pathlib import Path

import matplotlib.pyplot as plt
import seaborn as sns

# Dataset Exploration

For the final recommender we decided to upgrade the dataset to a more complete one. This is the data exploration of it.

Dataset Name: SkinSAFE Skincare Products (50,000+)

Kaggle Link: https://www.kaggle.com/datasets/kazireyazulhasan/skinsafe-skincare-db-50000-products?resource=download

In [43]:
BASE_DIR = Path().resolve() 
DATASET_PATH = BASE_DIR / ".." / ".." / "MiguelLib" / "datasets" / "data" / "products_raw.csv"
DATASET_PATH = DATASET_PATH.resolve()


df = pd.read_csv(DATASET_PATH)
df.head()

,product_name,brand,usage_type,category,ingredients,image_url,product_url
0,No7 Hydra Luminous Aqua Release Skin Perfector...,No7,Face,Tinted Moisturizers,"Propylene Glycol, Alcohol Denat, Anhydroxylito...",https://cdn1.skinsafeproducts.com/photo/E84797...,https://www.skinsafeproducts.com/no7-hydra-lum...
1,"Tower 28 SOS Daily Rescue Facial Spray, 4 fl o...",Tower 28,Face,Spray Moisturizer,"Hypochlorous Acid, Sodium Chloride, Water",https://cdn1.skinsafeproducts.com/photo/488026...,https://www.skinsafeproducts.com/tower-28-sos-...
2,"Clinique Takes the Day Off Cleansing Oil, 6.7 ...",Clinique,Facial Cleansers,Facial Cleansing Oil,"Bht, Caprylyl Glycol, Cetyl Ethylhexanoate, Gl...",https://cdn1.skinsafeproducts.com/photo/53EEB4...,https://www.skinsafeproducts.com/clinique-take...
3,"Necessaire The Body Lotion, Fragrance Free, 15...",Necessaire,Body,Lotions,"Aqua/Water/Eau, Caprylhydroxamic Acid, Capryli...",https://cdn1.skinsafeproducts.com/photo/D9ADD5...,https://www.skinsafeproducts.com/necessaire-th...
4,Paula’s Choice Resist Weightless Advanced Repa...,Paula's Choice,Toners & Astringents,Toners,"Acetyl Glucosamine, Adenosine, Butylene Glycol...",https://cdn1.skinsafeproducts.com/photo/2A9B12...,https://www.skinsafeproducts.com/paula-s-choic...


### Overview

In [44]:
print("=== Dataset Overview ===")
print(f"Number of rows: {df.shape[0]}")
print(f"Number of columns: {df.shape[1]}")
print("\nColumns:")
for col in df.columns:
    print(f" - {col} ({df[col].dtype})")

=== Dataset Overview ===
Number of rows: 50969
Number of columns: 7

Columns:
 - product_name (object)
 - brand (object)
 - usage_type (object)
 - category (object)
 - ingredients (object)
 - image_url (object)
 - product_url (object)


In [45]:
print("\n=== Missing Values ===")
missing_counts = df.isnull().sum()
missing_pct = (df.isnull().mean() * 100).round(2)
missing_df = pd.DataFrame({"count": missing_counts, "percent": missing_pct})
display(missing_df)

dupes = df.duplicated(subset=["product_name", "brand"]).sum()
print(f"\nDuplicate rows (product_name + brand): {dupes}")


=== Missing Values ===


,count,percent
product_name,0,0.0
brand,0,0.0
usage_type,0,0.0
category,0,0.0
ingredients,0,0.0
image_url,0,0.0
product_url,0,0.0



Duplicate rows (product_name + brand): 623


In [46]:
dupes = df.duplicated(subset=["product_name", "brand", "ingredients"]).sum()
print(f"\nDuplicate rows (product_name + brand + ingredients): {dupes}")


Duplicate rows (product_name + brand + ingredients): 622


In [ ]:
duplicate_rows = df[df.duplicated(subset=["product_name", "brand", "ingredients"], keep=False)]
duplicate_rows = duplicate_rows.sort_values(["product_name", "brand"]).reset_index(drop=True)
display(duplicate_rows.head(20))  

,product_name,brand,usage_type,category,ingredients,image_url,product_url
0,100% Pure Face: Moisturizer: Retinol Restorati...,100% Pure,Face,Nighttime Moisturizers,"African Shea Butter, Ascorbyl Palmitate (Vitam...",https://cdn1.skinsafeproducts.com/photo/557B89...,https://www.skinsafeproducts.com/100-pure-face...
1,100% Pure Face: Moisturizer: Retinol Restorati...,100% Pure,Face,Nighttime Moisturizers,"African Shea Butter, Ascorbyl Palmitate (Vitam...",https://cdn1.skinsafeproducts.com/photo/557B89...,https://www.skinsafeproducts.com/100-pure-face...
2,365 Everyday Value Sandalwood Foaming Hand Soa...,365 Everyday Value,Cleansers,Liquid or Cream Hand Soaps,"African Shea Butter, Allantoin, Aloe Barbadens...",https://cdn1.skinsafeproducts.com/photo/AFECD0...,https://www.skinsafeproducts.com/365-everyday-...
3,365 Everyday Value Sandalwood Foaming Hand Soa...,365 Everyday Value,Cleansers,Liquid or Cream Hand Soaps,"African Shea Butter, Allantoin, Aloe Barbadens...",https://cdn1.skinsafeproducts.com/photo/AFECD0...,https://www.skinsafeproducts.com/365-everyday-...
4,"40 Carrots Carrot + C Vitamin Serum, 1-Ounce B...",40 Carrots,Skin Treatments,Serums,"Ascorbic Acid (Vitamin C), Beta-Carotene, Dauc...",https://cdn1.skinsafeproducts.com/photo/D295EE...,https://www.skinsafeproducts.com/40-carrots-ca...
5,"40 Carrots Carrot + C Vitamin Serum, 1-Ounce B...",40 Carrots,Skin Treatments,Serums,"Ascorbic Acid (Vitamin C), Beta-Carotene, Dauc...",https://cdn1.skinsafeproducts.com/photo/D295EE...,https://www.skinsafeproducts.com/40-carrots-ca...
6,"A La Maison De Provence Hand & Body Soap, Oat ...",A La Maison de Provence,Cleansers,Bar Soaps,"African Shea Butter, Argania Spinosa (Argan) O...",https://cdn1.skinsafeproducts.com/photo/651AE0...,https://www.skinsafeproducts.com/a-la-maison-d...
7,"A La Maison De Provence Hand & Body Soap, Oat ...",A La Maison de Provence,Cleansers,Bar Soaps,"African Shea Butter, Argania Spinosa (Argan) O...",https://cdn1.skinsafeproducts.com/photo/651AE0...,https://www.skinsafeproducts.com/a-la-maison-d...
8,"A La Maison French Liquid Soap, White Tea, 16....",A La Maison de Provence,Cleansers,Liquid Cleansers & Soaps,"Argania Spinosa (Argan) Oil, Coconut Oil, Esse...",https://cdn1.skinsafeproducts.com/photo/i1EDr_...,https://www.skinsafeproducts.com/a-la-maison-f...
9,"A La Maison French Liquid Soap, White Tea, 16....",A La Maison de Provence,Cleansers,Liquid Cleansers & Soaps,"Argania Spinosa (Argan) Oil, Coconut Oil, Esse...",https://cdn1.skinsafeproducts.com/photo/i1EDr_...,https://www.skinsafeproducts.com/a-la-maison-f...


### Columns

#### Brand

In [50]:
print("\n--- Product Brands ---")
brand_counts = df["brand"].value_counts()
print(brand_counts)
print("\nNumber of unique brands:", df["brand"].nunique())


--- Product Brands ---
brand
Bath & Body Works    2072
Neutrogena            615
Olay                  545
Nivea                 442
Clinique              436
                     ... 
Uni.Q                   1
Evil Goods              1
Eco Max                 1
Petite Skin Co.         1
Manda Organic           1
Name: count, Length: 6959, dtype: int64

Number of unique brands: 6959


In [61]:
top_brands = df["brand"].value_counts().head(10)

#### Usage Type

In [55]:
print("\n--- Usage Type---")
usage_counts = df["usage_type"].value_counts()
print(usage_counts)
print("\nNumber of unique usage types:", df["usage_type"].nunique())


--- Usage Type---
usage_type
Body                               8745
Skin Treatments                    8125
Face                               7952
Cleansers                          7555
Facial Cleansers                   3605
Lip Care                           2947
Skin Care                          2308
Anti-Aging/Anti-Wrinkle            1567
Hand                               1489
Eye Cream, Gel, Oils, & Serum      1324
Exfoliators, Polishes, & Scrubs    1170
Toners & Astringents               1068
Moisturizers                        817
Cloths, Towelettes, & Wipes         808
Eyes                                467
Acids & Peels                       303
Alternative/Natural Skin Care       280
Pore Treatments                     175
Feet                                132
Beauty                              122
Bar Soaps                            10
Name: count, dtype: int64

Number of unique usage types: 21


#### Category

In [56]:
print("\n--- Product Categories ---")
category_counts = df["category"].value_counts()
print(category_counts)
print("\nNumber of unique categories:", df["category"].nunique())


--- Product Categories ---
category
Creams                          7388
Lotions                         5168
Serums                          4190
Facial Masks                    3413
Liquid or Cream Hand Soaps      2900
                                ... 
Complexes                          7
Vitamin E Capsules                 2
Dry Shower Foam                    2
Anti-Aging/Anti-Wrinkle (RX)       1
Teas                               1
Name: count, Length: 101, dtype: int64

Number of unique categories: 101


#### Ingredients

In [51]:
print("\n--- Ingredients Overview ---")
print(df["ingredients"].head(5))
ingredient_lengths = df["ingredients"].str.len()
print("\nIngredient text lengths stats:")
print(ingredient_lengths.describe())
print("\nRandom sample of ingredients:")
print(df["ingredients"].sample(10, random_state=42))


--- Ingredients Overview ---
0    Propylene Glycol, Alcohol Denat, Anhydroxylito...
1            Hypochlorous Acid, Sodium Chloride, Water
2    Bht, Caprylyl Glycol, Cetyl Ethylhexanoate, Gl...
3    Aqua/Water/Eau, Caprylhydroxamic Acid, Capryli...
4    Acetyl Glucosamine, Adenosine, Butylene Glycol...
Name: ingredients, dtype: object

Ingredient text lengths stats:
count    50969.000000
mean       519.474504
std        316.806039
min          4.000000
25%        301.000000
50%        468.000000
75%        678.000000
max       5044.000000
Name: ingredients, dtype: float64

Random sample of ingredients:
9613     Propylene Glycol, Beeswax, Benzophenone-3, But...
13462    Avena Sativa (Oat) Kernel Flour, Benzyl Alcoho...
27872    Bht, Fragrance, Isopropyl Myristate, Peg-40 So...
20087    Propylene Glycol, Alpha-Isomethyl Ionone, Aqua...
27767    Benzyl Alcohol, Canola Oil, Fragrance, Isoprop...
40503    Ammonium Acryloyldimethyltaurate/Vp Copolymer,...
32436    1,2-Hexanediol, Alcohol De

## Summary

**Dataset Overview**

- Total rows: 50,969

- Total columns: 7 – product_name, brand, usage_type, category, ingredients, image_url, product_url

- No missing values in any column.

**Duplicates**

- Duplicate rows by product_name + brand: 623

- Duplicate rows by product_name + brand + ingredients: 622

- Total duplicate entries: 1,242

- We will have to further review the duplicates to see if they are true duplicates to drop them.

**Brands**

- Total unique brands: 6,959

- Top brands by product count:

    - Bath & Body Works – 2,072

    - Neutrogena – 615

    - Olay – 545

    - Nivea – 442

    - Clinique – 436

**Usage Types**

- Total unique usage types: 21

- Most common usage types:

    - Body – 8,745

    - Skin Treatments – 8,125

    - Face – 7,952

    - Cleansers – 7,555

    - Facial Cleansers – 3,605

**Categories**

- Total unique categories: 101

- Most frequent categories:

    - Creams – 7,388

    - Lotions – 5,168

    - Serums – 4,190

    - Facial Masks – 3,413

    - Liquid or Cream Hand Soaps – 2,900

**Ingredients**

- Ingredients column is text-heavy, average length: ~519 characters

- Min length: 4, Max length: 5,044 characters

- Random samples show that ingredient lists do seem to follow INCI standard, but we will have to further inspect to make sure. 



**Important Notes**

- Duplicates have must be reviewed

- Ingredient lists must be deeply inspected to make sure they are usable 

- Prize is not a column and it is very relevant to the recommender 

## Why is this a good dataset to train our recommender with?

This dataset is highly suitable for our skincare recommender because it is comprehensive, covering 50,000+ products across hundreds of brands, categories, and usage types. It includes detailed ingredient lists in standard format, which allows us to match products to user preferences and skin concerns effectively. The dataset is clean and even though it has 600+ duplicates it is nothing that preprocessing can't handle. Most importantly, the ingredient lists are extensive and seem to follow the same format (INCI). This dataset provides the necessary resources to create a relevant skincare recommender that covers a range of products and preferences.